alacritty_terminal/src/tty/unix.rs

```diff
@@ -121,17 +118,30 @@ impl Pty {
     }
 }
 
-#[cfg(target_os = "macos")]
-fn default_shell(pw: &Passwd<'_>) -> Program {
-    let shell_name = pw.shell.rsplit('/').next().unwrap();
-    let argv = vec![String::from("-c"), format!("exec -a -{} {}", shell_name, pw.shell)];
-
-    Program::WithArgs { program: "/bin/bash".to_owned(), args: argv }
+fn default_shell(pw: &Passwd<'_>) -> String {
+    // Look for a shell in SHELL, then in passwd.
+    env::var("SHELL").unwrap_or_else(|_| pw.shell.to_owned())
 }
 
 #[cfg(not(target_os = "macos"))]
-fn default_shell(pw: &Passwd<'_>) -> Program {
-    Program::Just(env::var("SHELL").unwrap_or_else(|_| pw.shell.to_owned()))
+fn default_shell_command(pw: &Passwd<'_>) -> Command {
+    Command::new(default_shell(pw))
+}
+
+#[cfg(target_os = "macos")]
+fn default_shell_command(pw: &Passwd<'_>) -> Command {
+    // On macOS, use the `login` command to launch the shell so programs like `w`, `who`, and
+    // `logname` work.
+    let shell = default_shell(pw);
```

Comment:
This comment above the `default_shell` is somewhat confusing.

I'd recommend putting a newline between shell_name and login_command, then putting the comment above the actual `login` command spawn.